### PR TITLE
Hybrid offline mode

### PR DESCRIPTION
**Pull request by evgenyz**
This is an umbrella PR for hybrid offline mode support

- systemdunit* (+tests)
- selinuxboolean (+tests)
- selinuxsecuritycontext (+tests)
- process58 (+tests)
- environmentvariable58 (+tests)
- partition (+tests)

### CODE REVIEW

**Comment by evgenyz** [comment]
@openscap-ci test this, please

**Comment by jhrozek** [comment]
Awesome work. I tested the systemdservices one, it seems to work. Does does the selinuxboolean one. I tried testing the selinuxsecuritycontext one, but that one doesn't work. I don't know if I maybe did something wrong, so I'm going to attach a log. I don't know how to test the process58 one, it seems to only be mentioned in `shared/macros-oval.jinja`

**Comment by jhrozek** [comment]
But, this is a really good progress, thank you!

**Comment by jhrozek** [comment]
[selinux_devices.log](https://github.com/OpenSCAP/openscap/files/4361344/selinux_devices.log)

**Comment by evgenyz** [comment]
@openscap-ci test this, please

**Comment by evgenyz** [comment]
@jhrozek The selinuxsecuritycontext test is not executed or even mentioned in the log, I wonder why. Can you please also attach the DS?

**Comment by jhrozek** [comment]
Sure, attached
[ssg-ocp4-ds.xml.gz](https://github.com/OpenSCAP/openscap/files/4370402/ssg-ocp4-ds.xml.gz)

**Comment by jhrozek** [comment]
btw I tested with a DS build atop master with the following modifications:
```
diff --git a/linux_os/guide/system/selinux/selinux-booleans/sebool_fips_mode/rule.yml b/linux_os/guide/system/selinux/selinux-booleans/sebool_fips_mode/rule.yml
index db087e90e..af3b39930 100644
--- a/linux_os/guide/system/selinux/selinux-booleans/sebool_fips_mode/rule.yml
+++ b/linux_os/guide/system/selinux/selinux-booleans/sebool_fips_mode/rule.yml
@@ -1,6 +1,6 @@
 documentation_complete: true
 
-prodtype: rhel7,rhel8,rhv4
+prodtype: rhel7,rhel8,rhv4,ocp4
 
 title: 'Enable the fips_mode SELinux Boolean'
 
diff --git a/linux_os/guide/system/selinux/selinux_all_devicefiles_labeled/rule.yml b/linux_os/guide/system/selinux/selinux_all_devicefiles_labeled/rule.yml
index a274dbff1..a61ee3061 100644
--- a/linux_os/guide/system/selinux/selinux_all_devicefiles_labeled/rule.yml
+++ b/linux_os/guide/system/selinux/selinux_all_devicefiles_labeled/rule.yml
@@ -1,6 +1,6 @@
 documentation_complete: true
 
-prodtype: fedora,ol7,ol8,rhel6,rhel7,rhel8,rhv4,wrlinux1019
+prodtype: fedora,ol7,ol8,rhel6,rhel7,rhel8,rhv4,wrlinux1019,ocp4
 
 title: 'Ensure No Device Files are Unlabeled by SELinux'
 
diff --git a/ocp4/profiles/moderate.profile b/ocp4/profiles/moderate.profile
index 317a7b331..513101470 100644
--- a/ocp4/profiles/moderate.profile
+++ b/ocp4/profiles/moderate.profile
@@ -228,6 +228,7 @@ selections:
     - selinux_state
     - var_selinux_policy_name=targeted
     - selinux_policytype
+    - selinux_all_devicefiles_labeled
 
     ### Application Whitelisting (RHEL 8)
     - package_fapolicyd_installed
@@ -253,6 +254,7 @@ selections:
     - configure_openssl_crypto_policy
     - configure_kerberos_crypto_policy
     - enable_dracut_fips_module
+    - sebool_fips_mode
 
     #######################################################
     ### CONFIGURATION ANNEX TO THE PROTECTION PROFILE

```

**Comment by evgenyz** [comment]
From the DS: `          <ns10:Rule id="xccdf_org.ssgproject.content_rule_selinux_all_devicefiles_labeled" selected="false" severity="medium">`. I wonder why it is "not selected".

**Comment by evgenyz** [comment]
The most likely reason is `paltform: machine` in `linux_os/guide/system/selinux/group.yml`.

File: src/OVAL/probes/unix/linux/selinuxboolean_probe.c [hunk 1]
```diff
@@ -44,28 +44,154 @@
...
+	char path[PATH_MAX] = {0};
+	char *buf = NULL, *p;
+	FILE *fp = NULL;
+	size_t len;
+	ssize_t num;
+
+	if (!prefix)
+		return;
+
+	if (verify_selinuxmnt_prefixed(prefix, SELINUXMNT) == 0) return;
```
Comment (jan-cerny) [inline src/OVAL/probes/unix/linux/selinuxboolean_probe.c]: coding style!

File: src/OVAL/probes/unix/linux/selinuxboolean_probe.c [hunk 2]
```diff
@@ -44,28 +44,154 @@
...
+
+	if (verify_selinuxmnt_prefixed(prefix, OLDSELINUXMNT) == 0) return;
```
Comment (jan-cerny) [inline src/OVAL/probes/unix/linux/selinuxboolean_probe.c]: coding style!

File: src/OVAL/probes/unix/linux/selinuxboolean_probe.c [hunk 3]
```diff
@@ -44,28 +44,154 @@
...
+
+/* End of borrowing from SELinux */
+
 static int get_selinuxboolean(SEXP_t *ut_ent, probe_ctx *ctx)
 {
 	int err = 1, active, pending, len, i;
 	SEXP_t *boolean, *item;
 	char **booleans;
 
-	if ( ! is_selinux_enabled()) {
```
Comment (jan-cerny) [inline src/OVAL/probes/unix/linux/selinuxboolean_probe.c]: What will happen now if SELinux isn't enabled?

File: src/OVAL/probes/unix/linux/selinuxsecuritycontext_probe.c [hunk 4]
```diff
@@ -332,11 +334,11 @@ int selinuxsecuritycontext_probe_main(probe_ctx *ctx, void *arg)
...
-	behaviors = probe_obj_getent (probe_in, "behaviors", 1);
-	path      = probe_obj_getent (probe_in, "path",      1);
-	filename  = probe_obj_getent (probe_in, "filename",  1);
-	filepath  = probe_obj_getent (probe_in, "filepath", 1);
-	pid       = probe_obj_getent (probe_in, "pid", 1);
+	behaviors = probe_obj_getent(probe_in, "behaviors", 1);
+	path      = probe_obj_getent(probe_in, "path",      1);
+	filename  = probe_obj_getent(probe_in, "filename",  1);
+	filepath  = probe_obj_getent(probe_in, "filepath",  1);
+	pid       = probe_obj_getent(probe_in, "pid",       1);
```
Comment (jan-cerny) [inline src/OVAL/probes/unix/linux/selinuxsecuritycontext_probe.c]: please don't do only formatting changes

File: src/OVAL/probes/unix/process58_probe.c [hunk 5]
```diff
@@ -108,15 +108,15 @@ extern char const *_cap_names[];
...
+	unsigned pid;
+	unsigned ppid;
+	long priority;
 	int ruid;
-        const char *scheduling_class;
-        const char *start_time;
-        const char *tty;
+	const char *scheduling_class;
+	const char *start_time;
+	const char *tty;
```
Comment (jan-cerny) [inline src/OVAL/probes/unix/process58_probe.c]: please don't do only formatting changes

File: src/OVAL/probes/unix/linux/selinuxboolean_probe.c (hunk 3 above)
Comment (evgenyz) [reply src/OVAL/probes/unix/linux/selinuxboolean_probe.c]: SElinux might be disabled in the container where the scanner is running, but enabled on the host. In this situation we still can check SELinux booleans in `/host/sys/fs/selinux`.

Moreover I don't really understand why this check was needed in the first place. If there is no mounted selinux filesystem, the probe will return an error. If there is — it will fetch values.

**Comment by jhrozek** [comment]
So I ran a scan using a content with this diff atop the changes I did recently:
```
diff --git a/linux_os/guide/system/selinux/group.yml b/linux_os/guide/system/selinux/group.yml
index 6525cb491..e1863d4d0 100644
--- a/linux_os/guide/system/selinux/group.yml
+++ b/linux_os/guide/system/selinux/group.yml
@@ -29,5 +29,3 @@ description: |-
     {{% elif product == "ol7" %}}
     For more information on SELinux, see <b>{{{ weblink(link="https://docs.oracle.com/cd/E52668_01/E54669/html/ol7-s1-syssec.html") }}}</b>.
     {{% endif %}}
-
-platform: machine
```

But the result is pretty much the same. I'm going to attach the logs regardless.

**Comment by jhrozek** [comment]
[xccdf_org.ssgproject.content_rule_selinux_all_devicefiles_labeled.log](https://github.com/OpenSCAP/openscap/files/4375597/xccdf_org.ssgproject.content_rule_selinux_all_devicefiles_labeled.log)

**Comment by evgenyz** [comment]
Okay. Here is the thing. Your DS containt OCIL tests insted of OVAL. OpenSCAP does not support OCIL tests.

```xml
<ns10:check system="http://scap.nist.gov/schema/ocil/2">
<ns10:check-content-ref href="ssg-ocp4-ocil.xml" name="ocil:ssg-selinux_all_devicefiles_labeled_ocil:questionnaire:1"/>
</ns10:check>
```
How you managed to build it in that way — I don't know. @yuumasato what are we missing here?

**Comment by redhatrises** [comment]
> Okay. Here is the thing. Your DS contains OCIL tests instead of OVAL. OpenSCAP does not support OCIL tests.
> 
> ```
> <ns10:check system="http://scap.nist.gov/schema/ocil/2">
> <ns10:check-content-ref href="ssg-ocp4-ocil.xml" name="ocil:ssg-selinux_all_devicefiles_labeled_ocil:questionnaire:1"/>
> </ns10:check>
> ```
> 
> How you managed to build it in that way — I don't know. @yuumasato what are we missing here?

OCIL is generated by all datastreams that are built by us. I believe that the error in the log is a standard error.

**Comment by evgenyz** [comment]
Sorry, I meant that there is only OCIL test in that DS. Here is for example same thing from my Fedora's DS:

```xml
<ns10:check system="http://oval.mitre.org/XMLSchema/oval-definitions-5">
   <ns10:check-content-ref href="ssg-fedora-oval.xml" name="oval:ssg-selinux_all_devicefiles_labeled:def:1"/>
</ns10:check>
<ns10:check system="http://scap.nist.gov/schema/ocil/2">
   <ns10:check-content-ref href="ssg-fedora-ocil.xml" name="ocil:ssg-selinux_all_devicefiles_labeled_ocil:questionnaire:1"/>
</ns10:check>
```

**Comment by matusmarhefka** [comment]
@evgenyz I know that this is a draft so far, but have you consider adding also tests?

File: src/OVAL/probes/unix/linux/selinuxsecuritycontext_probe.c (hunk 4 above)
Comment (evgenyz) [reply src/OVAL/probes/unix/linux/selinuxsecuritycontext_probe.c]: So, when it is appropriate to fix such formatting error in your opinion?

**Comment by evgenyz** [comment]
@matusmarhefka Yes, I'm considering tests as well.

**Comment by jhrozek** [comment]
> Okay. Here is the thing. Your DS contains OCIL tests instead of OVAL. OpenSCAP does not support OCIL tests.
> 
> ```
> <ns10:check system="http://scap.nist.gov/schema/ocil/2">
> <ns10:check-content-ref href="ssg-ocp4-ocil.xml" name="ocil:ssg-selinux_all_devicefiles_labeled_ocil:questionnaire:1"/>
> </ns10:check>
> ```
> 
> How you managed to build it in that way — I don't know. @yuumasato what are we missing here?

I used the `./utils/build_ds_container.sh ocp4` script in the repo without any additional parameters. And other rules seem to work, so is that an issue only with those rules or that family of rules?

btw what do you think about submitting the systemd probe patches separately so they can be merged and we can make use of them?

**Comment by evgenyz** [comment]
@jhrozek I found the issue. You also have to add `        <platform>multi_platform_ocp4</platform>
` to `linux_os/guide/system/selinux/selinux_all_devicefiles_labeled/oval/shared.xml`.

**Comment by evgenyz** [comment]
@jhrozek the OVAL check of the rule was excluded from the DS because of this, and there was a warning about it during the build process: `WARNING: OVAL check 'selinux_all_devicefiles_labeled' was not found, removing <check-content> element from the XCCDF rule.`

**Comment by evgenyz** [comment]
> btw what do you think about submitting the systemd probe patches separately so they can be merged and we can make use of them?

I'm still trying to figure out how to test it in upstream test suite, it's not good to add such a change without even primitive smoke-test.

File: src/OVAL/probes/unix/process58_probe.c (hunk 5 above)
Comment (evgenyz) [reply src/OVAL/probes/unix/process58_probe.c]: And the same question here. When we are going to fix it? Asking again because this formatting error is different (could be mitigated semi-automatically by a code beautifier).

**Comment by jhrozek** [comment]
Thanks @evgenyz that helped! Also, today I learned something, I thought that the platforms are inherited from rule.yml.
Thank you, is there anything else you would like me to test?

**Comment by evgenyz** [comment]
@jhrozek It would be great if you could confirm that `environmentvariable58` and `partition` probes are working correctly in operator as well.

**Comment by evgenyz** [comment]
@matusmarhefka Tests are there, although offline ones are pretty naive. To make them more realistic we have to utilise podman in our test suite, and do it a very perculiar way, with systemd hacks and privileged tricks. And I was not able to produce such a setup yet, even for myself.

@jan-cerny PR is complete, I won't add any new probes here, only fixes. I'm switching it from draft. Can you please also react on my questions in your former review?

@jhrozek I'd would be great you you could conduct ad cumulative test of this PR in the cluster before we land it into maint-1.3.

File: tests/probes/environmentvariable58/test_probes_environmentvariable58_offline_mode.sh [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    grep -Ei "(W:|Entity has no value)" $stderr && ret_val=1
+
+    rm $stderr
+
+    rm -rf ${tmpdir}
+
+    return $ret_val
+}
+
+test_probes_environmentvariable58_offline_mode
```
Comment (matusmarhefka) [inline tests/probes/environmentvariable58/test_probes_environmentvariable58_offline_mode.sh]: I suggest using our shared `test_` functions here, we have it documented in https://github.com/OpenSCAP/openscap/blob/maint-1.3/docs/contribute/testing.adoc and as an example you can take a look at https://github.com/OpenSCAP/openscap/blob/maint-1.3/tests/offline_mode/test_offline_mode_textfilecontent54.sh#L52

It makes the verbose output more readable and it is helpful when one needs to debug some issues. This also applies for the other tests.

File: tests/probes/partition/test_probes_partition_offline_mode.sh [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    local RF="test_probes_partition_offline_mode.results.xml"
+    echo "result file: $RF"
+    local stderr=$(mktemp $1.err.XXXXXX)
+    echo "stderr file: $stderr"
+
+    [ -f $RF ] && rm -f $RF
+
+    tmpdir=$(mktemp -t -d "test_offline_mode_partition.XXXXXX")
+
+    ln -s /proc $tmpdir
```
Comment (matusmarhefka) [inline tests/probes/partition/test_probes_partition_offline_mode.sh]: Does it make sense to parametrize this function and try it also on empty filesystem? just to verify if probe can handle it?

File: tests/probes/process58/test_probes_process58_offline_mode.sh [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    local RF="test_probes_process58_offline_mode.results.xml"
+    echo "result file: $RF"
+    local stderr=$(mktemp $1.err.XXXXXX)
+    echo "stderr file: $stderr"
+
+    [ -f $RF ] && rm -f $RF
+
+    tmpdir=$(mktemp -t -d "test_offline_mode_process58.XXXXXX")
+
+    ln -s /proc $tmpdir
```
Comment (matusmarhefka) [inline tests/probes/process58/test_probes_process58_offline_mode.sh]: Same here and probably also in other offline mode tests - maybe we can test empty filesystem which emulates missing /proc directory, probe should be able to handle it

File: utils/oscap-podman [hunk 9]
```diff
@@ -99,9 +99,8 @@ if [ ! -f "$DIR/run/.containerenv" ]; then
     touch "$DIR/run/.containerenv"
 fi
 
-for VAR in `podman inspect $ID --format '{{join .Config.Env " "}}'`; do
-    eval "export OSCAP_OFFLINE_$VAR"
-done
+export OSCAP_CONTAINER_VARS
+OSCAP_CONTAINER_VARS=`podman inspect $ID --format '{{join .Config.Env "\n"}}'`
```
Comment (matusmarhefka) [inline utils/oscap-podman]: What is the reason of this change? I am not against, just asking.

Comment (evgenyz) [reply utils/oscap-podman]: The idea is to create an independent code branch for consuming sideline source of environment variables instead of polluting process variables when we are in offline mode but don't have a running system. This is kind of special things among probes and hopefully it will evolve with introduction of system meta information.

Comment (evgenyz) [reply utils/oscap-podman]: See https://github.com/OpenSCAP/openscap/pull/1499/files#diff-c414bc5e3aa458311c624f11370a735fR107

Comment (matusmarhefka) [reply utils/oscap-podman]: Thanks, makes sense.

File: src/OVAL/probes/unix/process58_probe.c (hunk 5 above)
Comment (jan-cerny) [reply src/OVAL/probes/unix/process58_probe.c]: Probably never, only if we need eg. rename these variables or refactor the code.

We enforce the coding style only on new or edited code. This is because people often apply their patches on top of our code and they will have to rework them if you change the indents or code style.

File: tests/probes/partition/test_probes_partition_offline_mode.sh (hunk 7 above)
Comment (evgenyz) [reply tests/probes/partition/test_probes_partition_offline_mode.sh]: Done, except for environmentvariable58, where it does not make any sense.

File: tests/probes/environmentvariable58/test_probes_environmentvariable58_offline_mode.sh (hunk 6 above)
Comment (evgenyz) [reply tests/probes/environmentvariable58/test_probes_environmentvariable58_offline_mode.sh]: Fixed

File: src/OVAL/probes/unix/process58_probe.c (hunk 5 above)
Comment (evgenyz) [reply src/OVAL/probes/unix/process58_probe.c]: Reverted.

File: src/OVAL/probes/unix/linux/selinuxsecuritycontext_probe.c (hunk 4 above)
Comment (evgenyz) [reply src/OVAL/probes/unix/linux/selinuxsecuritycontext_probe.c]: Reverted.

File: tests/probes/selinuxsecuritycontext/test_probes_selinuxsecuritycontext_offline_mode.xml [hunk 10]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0"?>
```
Comment (matusmarhefka) [inline tests/probes/selinuxsecuritycontext/test_probes_selinuxsecuritycontext_offline_mode.xml]: Is there any differences compared to the `test_probes_selinuxsecuritycontext.xml`? If not, I would reuse it for offline test and remove this file.

File: tests/probes/systemdunitdependency/test_probes_systemdunitdependency_offline_mode.sh [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+    tmpdir=$(mktemp -t -d "test_offline_mode_systemdunitdependency.XXXXXX")
+
+    if [[ "${VALID}" == "true" ]]; then
+        ln -s /run $tmpdir
+    fi
+
+    set_chroot_offline_test_mode "$tmpdir"
+
+    $OSCAP oval eval --results $RF $DF
```
Comment (matusmarhefka) [inline tests/probes/systemdunitdependency/test_probes_systemdunitdependency_offline_mode.sh]: Check stderr.

File: tests/probes/systemdunitproperty/test_probes_systemdunitproperty_offline_mode.sh [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+    tmpdir=$(mktemp -t -d "test_offline_mode_systemdunitproperty.XXXXXX")
+
+    if [[ "${VALID}" == "true" ]]; then
+        ln -s /run $tmpdir
+    fi
+
+    set_chroot_offline_test_mode "$tmpdir"
+
+    $OSCAP oval eval --results $RF $DF
```
Comment (matusmarhefka) [inline tests/probes/systemdunitproperty/test_probes_systemdunitproperty_offline_mode.sh]: Check stderr.

File: src/OVAL/probes/unix/linux/systemdshared.h [hunk 13]
```diff
@@ -329,6 +331,15 @@ static DBusConnection *connect_dbus()
 	DBusError err;
 	dbus_error_init(&err);
 
+	const char *prefix = getenv("OSCAP_PROBE_ROOT");
+	if (prefix != NULL) {
+		char dbus_address[PATH_MAX] = {0};
+		snprintf(dbus_address, PATH_MAX, "unix:path=%s/run/dbus/system_bus_socket", prefix);
+		setenv("DBUS_SYSTEM_BUS_ADDRESS", dbus_address, 0);
```
Comment (matusmarhefka) [inline src/OVAL/probes/unix/linux/systemdshared.h]: So systemd probes collect all data only through DBus?

File: src/OVAL/probes/unix/linux/systemdunitdependency_probe.c [hunk 14]
```diff
@@ -232,7 +239,7 @@ int systemdunitdependency_probe_main(probe_ctx *ctx, void *probe_arg)
 	if (dbus_conn == NULL) {
 		dbus_error_free(&dbus_error);
 		SEXP_t *msg = probe_msg_creat(OVAL_MESSAGE_LEVEL_INFO, "DBus connection failed, could not identify systemd units.");
-		probe_cobj_set_flag(probe_ctx_getresult(ctx), SYSCHAR_FLAG_ERROR);
+		probe_cobj_set_flag(probe_ctx_getresult(ctx), ctx->offline_mode == PROBE_OFFLINE_NONE ? SYSCHAR_FLAG_ERROR : SYSCHAR_FLAG_NOT_COLLECTED);
```
Comment (matusmarhefka) [inline src/OVAL/probes/unix/linux/systemdunitdependency_probe.c]: Will this produce unknown rule result in case of `PROBE_OFFLINE_OWN` and failing to connect to DBus?

File: src/OVAL/probes/unix/linux/systemdshared.h (hunk 13 above)
Comment (evgenyz) [reply src/OVAL/probes/unix/linux/systemdshared.h:338]: Yes

**Comment by jhrozek** [comment]
Testing will cumulative patches:
 * systemdunit works. Tested by disabling audit and running `xccdf_org.ssgproject.content_rule_service_auditd_enabled`. The reverse, when the service is running passes.
 * selinuxboolean works. Tested by disabling `fips_mode` and running `xccdf_org.ssgproject.content_rule_sebool_fips_mode`. The reverse, when the boolean is enabled, passes.

However, `selinuxsecuritycontext` does not seem to be working. I mislabeled a file:
```
ls -lZ /dev/usbmon0
crw-------. 1 root root system_u:object_r:tmp_t:s0 244, 0 Apr  1 08:10 /dev/usbmon0
```
then I ran the `xccdf_org.ssgproject.content_rule_selinux_all_devicefiles_labeled` rule but I'm still getting a pass. I'm going to add a log of the test run.

Also, `process58` doesn't seem to be working. I'm getting `Requested offline mode is not supported by process58 probe.` I will also attach the full logs.

The `environmentvariable58` probe is also giving me strange results, it runs, but the rule still fails even though it should not. Again, logs are coming.
[xccdf_org.ssgproject.content_rule_root_path_no_dot.log](https://github.com/OpenSCAP/openscap/files/4414115/xccdf_org.ssgproject.content_rule_root_path_no_dot.log)
[xccdf_org.ssgproject.content_rule_etcd_cert_file.log](https://github.com/OpenSCAP/openscap/files/4414116/xccdf_org.ssgproject.content_rule_etcd_cert_file.log)
[xccdf_org.ssgproject.content_rule_selinux_all_devicefiles_labeled.log](https://github.com/OpenSCAP/openscap/files/4414117/xccdf_org.ssgproject.content_rule_selinux_all_devicefiles_labeled.log)

**Comment by evgenyz** [comment]
@jhrozek Regarding `environmentvariable58`. The object in the rule defined like this:
```xml
 <ind:environmentvariable58_object id="object_root_path_no_dot" version="1">
    <ind:pid xsi:nil="true" datatype="int" />
    <ind:name>PATH</ind:name>
  </ind:environmentvariable58_object>
```
which means that the probe should check environment variables of the scanner's process (because of `xsi:nil="true"` in **pid**). But that is just impossible due to the fact that there is no scanner process on the host. I'm not sure what to do here yet.

**Comment by evgenyz** [comment]
@jhrozek Did you test `process58` with the latest version of this branch?

**Comment by jhrozek** [comment]
On Wed, Apr 01, 2020 at 06:40:38AM -0700, Evgeny Kolesnikov wrote:
> @jhrozek Regarding `environmentvariable58`. The object in the rule defined like this:
> ```xml
>  <ind:environmentvariable58_object id="object_root_path_no_dot" version="1">
>     <ind:pid xsi:nil="true" datatype="int" />
>     <ind:name>PATH</ind:name>
>   </ind:environmentvariable58_object>
> ```
> which means that the probe should check environment variables of the scanner's process (because of `xsi:nil="true"` in **pid**). But that is just impossible due to the fact that there is no scanner process on the host. I'm not sure what to do here yet.

My suggestion would be to merge those probes that work as appropriate.
That would already help us.

**Comment by jhrozek** [comment]
On Wed, Apr 01, 2020 at 01:46:45PM +0000, Evgeny Kolesnikov wrote:
> @jhrozek Did you test `process58` with the latest version of this branch?

I am reasonably sure I did, but let me re-run the check again tomorrow.

(sorry, today's cluster is already gone and it would take long to
install a new one).

File: src/OVAL/probes/unix/linux/systemdunitdependency_probe.c (hunk 14 above)
Comment (evgenyz) [reply src/OVAL/probes/unix/linux/systemdunitdependency_probe.c:242]: It will produce unknown result. But I'm not sure if it's the desired behaviour in case it will be combined. See test `oval:0:tst:3` outcome in offline test.

File: tests/probes/selinuxsecuritycontext/test_probes_selinuxsecuritycontext_offline_mode.xml (hunk 10 above)
Comment (evgenyz) [reply tests/probes/selinuxsecuritycontext/test_probes_selinuxsecuritycontext_offline_mode.xml:1]: Fixed

File: tests/probes/partition/test_probes_partition.sh [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    $OSCAP oval eval --results $RF $DF 2> $stderr
+
+    if [ -f $RF ]; then
+        verify_results "def" $DF $RF 1 && verify_results "tst" $DF $RF 1
+        ret_val=$?
+    else
+        ret_val=1
+    fi
+
+    rm $stderr
```
[the remark this answers is not in the document]
Comment (evgenyz) [reply tests/probes/partition/test_probes_partition.sh:28]: Done

File: tests/probes/partition/test_probes_partition_offline_mode.sh [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+    tmpdir=$(mktemp -t -d "test_offline_mode_partition.XXXXXX")
+
+    if [[ "${VALID}" == "true" ]]; then
+        ln -s /proc $tmpdir
+    fi
+
+    set_chroot_offline_test_mode "$tmpdir"
+
+    $OSCAP oval eval --results $RF $DF
```
[the remark this answers is not in the document]
Comment (evgenyz) [reply tests/probes/partition/test_probes_partition_offline_mode.sh:26]: Done

File: tests/probes/process58/test_probes_process58_offline_mode.sh [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+
+    tmpdir=$(mktemp -t -d "test_offline_mode_process58.XXXXXX")
+
+    if [[ "${VALID}" == "true" ]]; then
+        ln -s /proc $tmpdir
+    fi
+
+    set_chroot_offline_test_mode "$tmpdir"
+
+    $OSCAP oval eval --results $RF $DF
```
[the remark this answers is not in the document]
Comment (evgenyz) [reply tests/probes/process58/test_probes_process58_offline_mode.sh:28]: Done

File: tests/probes/selinuxboolean/test_probes_selinuxboolean_offline_mode.sh [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+
+    tmpdir=$(mktemp -t -d "test_offline_mode_selinuxboolean.XXXXXX")
+
+    if [[ "${VALID}" == "true" ]]; then
+        ln -s /sys $tmpdir
+    fi
+
+    set_chroot_offline_test_mode "$tmpdir"
+
+    $OSCAP oval eval --results $RF $DF
```
[the remark this answers is not in the document]
Comment (evgenyz) [reply tests/probes/selinuxboolean/test_probes_selinuxboolean_offline_mode.sh:28]: Done

File: tests/probes/selinuxsecuritycontext/test_probes_selinuxsecuritycontext.sh [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+
+    probecheck "selinuxboolean" || return 255
+
+    local ret_val=0;
+    local DF="${srcdir}/test_probes_selinuxsecuritycontext.xml"
+    local RF="test_probes_selinuxsecuritycontext.results.xml"
+
+    [ -f $RF ] && rm -f $RF
+
+    $OSCAP oval eval --results $RF $DF
```
[the remark this answers is not in the document]
Comment (evgenyz) [reply tests/probes/selinuxsecuritycontext/test_probes_selinuxsecuritycontext.sh:17]: Done

File: tests/probes/selinuxsecuritycontext/test_probes_selinuxsecuritycontext_offline_mode.sh [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+
+    tmpdir=$(mktemp -t -d "test_offline_mode_selinuxsecuritycontext.XXXXXX")
+
+    if [[ "${VALID}" == "true" ]]; then
+        ln -s /dev $tmpdir
+    fi
+
+    set_chroot_offline_test_mode "$tmpdir"
+
+    $OSCAP oval eval --results $RF $DF
```
[the remark this answers is not in the document]
Comment (evgenyz) [reply tests/probes/selinuxsecuritycontext/test_probes_selinuxsecuritycontext_offline_mode.sh:26]: Done

File: tests/probes/systemdunitdependency/test_probes_systemdunitdependency_offline_mode.sh (hunk 11 above)
Comment (evgenyz) [reply tests/probes/systemdunitdependency/test_probes_systemdunitdependency_offline_mode.sh:32]: Done

File: tests/probes/systemdunitproperty/test_probes_systemdunitproperty_offline_mode.sh (hunk 12 above)
Comment (evgenyz) [reply tests/probes/systemdunitproperty/test_probes_systemdunitproperty_offline_mode.sh:31]: Done

**Comment by evgenyz** [comment]
@jhrozek The problem with `selinuxsecuritycontext` has been fixed.

**Comment by matusmarhefka** [comment]
I believe that all found issues have been resolved and there is no more feedback so I am merging this. Thank you @evgenyz for all the improvements in offline mode and also for adding tests.
